src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/PoolingAsyncValueTaskMethodBuilderT.cs

```diff
@@ -284,114 +285,85 @@ private sealed class StateMachineBox<TStateMachine> :
                 _valueTaskSource.Reset();
 
-                // If reusing the object would result in potentially wrapping around its version number, just throw it away.
-                // This provides a modicum of additional safety when ValueTasks are misused (helping to avoid the case where
-                // a ValueTask is illegally re-awaited and happens to do so at exactly 2^16 uses later on this exact same instance),
-                // at the expense of potentially incurring an additional allocation every 65K uses.
-                if ((ushort)_valueTaskSource.Version == ushort.MaxValue)
-                {
-                    return;
-                }
-
-                // If the thread static cache is empty, store this into it and bail.
+                // If the per-thread cache is empty, store this into it..
                 if (t_tlsCache is null)
                 {
                     t_tlsCache = this;
-                    return;
                 }
-
-                // Try to acquire the cache lock.  If there's any contention, or if the cache is full, we just throw away the object.
-                if (Interlocked.Exchange(ref s_cacheLock, 1) == 0)
+                else
                 {
-                    if (s_cacheSize < PoolingAsyncValueTaskMethodBuilder.s_valueTaskPoolingCacheSize)
+                    // Otherwise, store it into the per-core cache.
+                    ref StateMachineBox<TStateMachine>? slot = ref PerCoreCacheSlot;
+                    if (slot is null)
                     {
-                        // Push the box onto the cache stack for subsequent reuse.
-                        _next = s_cache;
-                        s_cache = this;
-                        s_cacheSize++;
-                        Debug.Assert(s_cacheSize > 0 && s_cacheSize <= PoolingAsyncValueTaskMethodBuilder.s_valueTaskPoolingCacheSize, "Expected cache size to be within bounds.");
+                        // Try to avoid the write if we know the slot isn't empty (we may still have a benign race condition and
+                        // overwrite what's there if something arrived in the interim).
+                        Volatile.Write(ref slot, this);
                     }
+                }
+            }
 
-                    // Release the lock.
-                    Volatile.Write(ref s_cacheLock, 0);
+            /// <summary>Gets the slot in <see cref="s_perCoreCache"/> for the current core.</summary>
+            private static ref StateMachineBox<TStateMachine>? PerCoreCacheSlot
+            {
+                [MethodImpl(MethodImplOptions.AggressiveInlining)] // only two callers are RentFrom/ReturnToCache
+                get
+                {
+                    // Get the current processor ID.  We need to ensure it fits within s_perCoreCache, so we
+                    // could % by its length, but we can do so instead by s_perCoreCache, which will be a const
```

Comment:
did you mean `Environment.ProcessorCount`?

```suggestion
                    // could % by its length, but we can do so instead by Environment.ProcessorCount, which will be a const
```

Reply:
Yup, thanks.